### PR TITLE
Add Tian Parametrization to Reactive Fluid Transport

### PR DESCRIPTION
**Pull request by danieldouglas92**
Implement a solid-fluid reaction scheme that uses parametrized phase diagrams from Tian et al., 2019 https://doi.org/10.1029/2019GC008488. 

A comparison between the peridotite phase diagram defined in the study and the ASPECT reproduction are shown below. Note that the y-axis is inverted in the ASPECT models since pressure increases with depth. The low-res output shows what is implemented as a test, and the high-res output is shown for better comparison to the study.

image

image

image

Pull Request Checklist. Please read and check each box with an X. Delete any part not applicable. Ask on the [forum](https://community.geodynamics.org/c/aspect) if you need help with any step.

*Describe what you did in this PR and why you did it.*

### Before your first pull request:

* [x] I have read the guidelines in our [CONTRIBUTING.md](../blob/main/CONTRIBUTING.md) document.

### For all pull requests:

* [x] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](https://aspect-documentation.readthedocs.io/en/latest/user/extending/testing/writing-tests.html) for the new feature/benchmark in the [tests/](../blob/main/tests/) directory.
* [x] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

### CODE REVIEW

**Review by naliboff — COMMENTED** [review]
@danieldouglas92 - Thanks for the contribution and at first glance the PR is in good shape. I made comments on the portions specific to the Tian et al. implementation. For now, a number of comments about documentation. Let's discuss how to proceed after #5245 is ready to be merged.

File: include/aspect/material_model/reactive_fluid_transport.h [hunk 1]
```diff
@@ -137,6 +141,10 @@ namespace aspect
 
         // Time scale for fluid release and absorption.
         double fluid_reaction_time_scale;
+        double max_peridotite_water;
+        double max_gabbro_water;
```
Comment (naliboff) [inline include/aspect/material_model/reactive_fluid_transport.h:145]: Look farther ahead, it looks like this and the other variables below are currently not being used to enforce upper limits? Is the plan to add them in a forthcoming commit?

**Review by naliboff — CHANGES_REQUESTED** [review]
@danieldouglas92 - Thanks for continuing the work on this and including all of the tests! Overall, I think the structure of the code is in good shape and the PR is not far off from being  after some moderate revisions.

I have some initial comments throughout, but one thing that stood out is that some of the code in your commits is not related to the Tian approximation and may have been introduced during a rebase?

File: include/aspect/material_model/reactive_fluid_transport.h [hunk 2]
```diff
@@ -137,6 +141,10 @@ namespace aspect
 
         // Time scale for fluid release and absorption.
         double fluid_reaction_time_scale;
+        double max_peridotite_water;
```
[the remark this answers is not in the document]
Comment (danieldouglas92) [reply include/aspect/material_model/reactive_fluid_transport.h:144]: I suspect that at least for now they're specific to the Tian scheme

File: source/material_model/reactive_fluid_transport.cc [hunk 3]
```diff
@@ -72,21 +161,43 @@ namespace aspect
               }
               case zero_solubility:
               {
-                // The fluid volume fraction in equilibrium with the solid
-                // at any point (stored in the melt_fractions vector) is
-                // equal to the sum of the bound fluid content and porosity.
-                const unsigned int bound_fluid_idx = this->introspection().compositional_index_for_name("bound_fluid");
+                // A fluid-rock reaction model where no reactions occur.
```
[the remark this answers is not in the document]
Comment (danieldouglas92) [reply source/material_model/reactive_fluid_transport.cc:164]: Ahh I must have made a mistake resolving the merge conflicts! good catch

File: source/material_model/reactive_fluid_transport.cc [hunk 4]
```diff
@@ -59,11 +147,12 @@ namespace aspect
       for (unsigned int q=0; q<in.temperature.size(); ++q)
         {
           const unsigned int porosity_idx = this->introspection().compositional_index_for_name("porosity");
-
+          const unsigned int bound_fluid_idx = this->introspection().compositional_index_for_name("bound_fluid");
```
[the remark this answers is not in the document]
Comment (danieldouglas92) [reply source/material_model/reactive_fluid_transport.cc:150]: `bound_fluid_idx` is also used in the `zero_solubility` scheme

File: source/material_model/reactive_fluid_transport.cc [hunk 5]
```diff
@@ -72,21 +161,43 @@ namespace aspect
...
+                tracked_rock_compositions.push_back(in.composition[q][peridotite_idx]);
+                tracked_rock_compositions.push_back(in.composition[q][gabbro_idx]);
+                tracked_rock_compositions.push_back(in.composition[q][MORB_idx]);
+                tracked_rock_compositions.push_back(in.composition[q][sediment_idx]);
+
+                std::vector<double> tian_eq_bound_water_content = tian_equilibrium_bound_water_content(in, q);
+
+                double average_eq_bound_water_content = MaterialUtilities::average_value (tracked_rock_compositions, tian_eq_bound_water_content, MaterialUtilities::arithmetic);
+
+                melt_fractions[q] = std::max(in.composition[q][bound_fluid_idx] + in.composition[q][porosity_idx] - average_eq_bound_water_content, 0.0);
```
[the remark this answers is not in the document]
Comment (danieldouglas92) [reply source/material_model/reactive_fluid_transport.cc:192]: `in.composition[q][bound_fluid_idx]` is the current bound fluid in the model, `tian_eq_bound_water_content` is the calculated bound water content for each of the 4 rock types, and `average_eq_bound_water_content` is the average of these 4 values

File: include/aspect/material_model/reactive_fluid_transport.h (hunk 1 above)
Comment (danieldouglas92) [reply include/aspect/material_model/reactive_fluid_transport.h:145]: These are used in the function `tian_equilibrium_bound_water_content()` to limit the bound water for the respective rock type.

File: source/material_model/reactive_fluid_transport.cc [hunk 6]
```diff
@@ -246,12 +356,26 @@ namespace aspect
                              "computed. If the model does not use operator splitting, this parameter is not used. "
                              "Units: yr or s, depending on the ``Use years "
                              "in output instead of seconds'' parameter.");
+          prm.declare_entry ("Maximum weight percent sediment", "3",
```
[the remark this answers is not in the document]
Comment (danieldouglas92) [reply source/material_model/reactive_fluid_transport.cc:359]: Yes this is already done in the function `tian_equilibrium_bound_water_content`

File: source/material_model/reactive_fluid_transport.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (naliboff) [reply source/material_model/reactive_fluid_transport.cc:195]: Thanks, I've add a suggested comment that describes this.

File: source/material_model/reactive_fluid_transport.cc [hunk 7]
```diff
@@ -75,18 +165,42 @@ namespace aspect
...
+                tracked_rock_mass_fractions.push_back(in.composition[q][MORB_idx]);
+                tracked_rock_mass_fractions.push_back(in.composition[q][sediment_idx]);
+
+                // The bound water content (water within the solid phase) for the four different rock types
+                std::vector<double> tian_eq_bound_water_content = tian_equilibrium_bound_water_content(in, q);
+
+                // average the water content between the four different rock types
+                double average_eq_bound_water_content = MaterialUtilities::average_value (tracked_rock_mass_fractions, tian_eq_bound_water_content, MaterialUtilities::arithmetic);
+
+                melt_fractions[q] = std::max(in.composition[q][bound_fluid_idx] + in.composition[q][porosity_idx] - average_eq_bound_water_content, 0.0);
```
Comment (naliboff) [inline source/material_model/reactive_fluid_transport.cc:195]:
```suggestion
                // The fluid volume fraction (stored in the melt_fractions vector) is equal to the porosity and the change
                // in bound fluid content (current bound fluid - updated average bound fluid).
                melt_fractions[q] = std::max(in.composition[q][bound_fluid_idx] + in.composition[q][porosity_idx] - average_eq_bound_water_content, 0.0);
```

File: source/material_model/reactive_fluid_transport.cc [hunk 8]
```diff
@@ -338,12 +497,6 @@ namespace aspect
       // After parsing the parameters for this model, parse parameters related to the base model.
       base_model->parse_parameters(prm);
       this->model_dependence = base_model->get_model_dependence();
-
-      if (fluid_solid_reaction_scheme == zero_solubility)
-        {
-          AssertThrow(this->get_material_model().is_compressible() == false,
-                      ExcMessage("The Fluid-reaction scheme zero solubility must be used with an incompressible base model."));
-        }
```
Comment (naliboff) [inline source/material_model/reactive_fluid_transport.cc:346]: Why is this removed?

File: tests/tian_MORB.prm [hunk 9]
```diff
@@ -0,0 +1,25 @@
+# This model generates a representation of a phase diagram for the equilibrium
+# wt% water that can be stored in an upper mantle peridotite (PUM) as shown in
+# Figure 11 of Tian et. al., 2019. This reproduces that phase diagram, with an important
```
Comment (naliboff) [inline tests/tian_MORB.prm:3]:
```suggestion
# Figure 11 of Tian et. al., 2019. This model reproduces that phase diagram, with an important
```

File: tests/tian_peridotite.prm [hunk 10]
```diff
@@ -0,0 +1,129 @@
+# This model generates a representation of a phase diagram for the equilibrium
+# wt% water that can be stored in an upper mantle peridotite (PUM) as shown in
+# Figure 11 of Tian et. al., 2019. This reproduces that phase diagram, with an important
```
Comment (naliboff) [inline tests/tian_peridotite.prm:3]:
```suggestion
# Figure 11 of Tian et. al., 2019. This model reproduces that phase diagram, with an important
```

File: tests/tian_peridotite.prm [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+  set Include melt transport                  = true
+end
+
+subsection Mesh refinement
+  set Initial global refinement               = 2
+end
+
+# Post processing
+subsection Postprocess
+  set List of postprocessors = visualization, composition statistics, velocity statistics
```
Comment (naliboff) [inline tests/tian_peridotite.prm:123]: The visualization files are included in this PR, but would be great to have so the data to reproduce the figures can be quickly plotted without running the models. Also, I would output in `ascii` rather than `vtu` format.

File: tests/tian_peridotite.prm [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+    set Thermal expansion coefficient = 0.0
+  end
+end
+
+subsection Melt settings
+  set Include melt transport                  = true
+end
+
+subsection Mesh refinement
+  set Initial global refinement               = 2
```
Comment (naliboff) [inline tests/tian_peridotite.prm:118]: This is fine for the test, but the resulting data set will be very coarse relative to the data in the Tian et al. figures. Can you make a note of this in the comments at the top of each PRM?

**Review by naliboff — COMMENTED** [review]
@danieldouglas92 - Thanks for the updates and this is getting much closer. I added another round of comments, and after that will probably only need to look at it once more (@jdannberg - can you also take a look at some point)? Thanks!

File: source/material_model/reactive_fluid_transport.cc (hunk 8 above)
Comment (danieldouglas92) [reply source/material_model/reactive_fluid_transport.cc:346]: A mistake!

File: tests/tian_peridotite.prm (hunk 11 above)
Comment (danieldouglas92) [reply tests/tian_peridotite.prm:123]: I actually didn't mean to include the .vtu files, if you think it would be worth it I can output it to ASCII but I don't think it's necessary for the tests to be effective so I can just remove the output

File: source/material_model/reactive_fluid_transport.cc [hunk 13]
```diff
@@ -50,6 +50,96 @@ namespace aspect
...
+                                                            LR_MORB_poly_coeffs, LR_sediment_poly_coeffs
+                                                           };
+
+      std::vector<std::vector<double>> csat_all_poly_coeffs {csat_peridotite_poly_coeffs, csat_gabbro_poly_coeffs, \
+                                                              csat_MORB_poly_coeffs, csat_sediment_poly_coeffs
+                                                             };
+
+      std::vector<std::vector<double>> Td_all_poly_coeffs {Td_peridotite_poly_coeffs, Td_gabbro_poly_coeffs, \
+                                                            Td_MORB_poly_coeffs, Td_sediment_poly_coeffs
+                                                           };
```
Comment (jdannberg) [inline source/material_model/reactive_fluid_transport.cc:94]: Is there a  reason you create these vectors here rather than in the .h file? This means that every time you call the function (in other words, for each cell in the model), they will be created and destroyed again.

File: source/material_model/reactive_fluid_transport.cc [hunk 14]
```diff
@@ -50,6 +50,96 @@ namespace aspect
...
+
+      std::vector<std::vector<double>> LR_all_poly_coeffs {LR_peridotite_poly_coeffs, LR_gabbro_poly_coeffs, \
+                                                            LR_MORB_poly_coeffs, LR_sediment_poly_coeffs
+                                                           };
+
+      std::vector<std::vector<double>> csat_all_poly_coeffs {csat_peridotite_poly_coeffs, csat_gabbro_poly_coeffs, \
+                                                              csat_MORB_poly_coeffs, csat_sediment_poly_coeffs
+                                                             };
+
+      std::vector<std::vector<double>> Td_all_poly_coeffs {Td_peridotite_poly_coeffs, Td_gabbro_poly_coeffs, \
```
Comment (jdannberg) [inline source/material_model/reactive_fluid_transport.cc:92]: I would give a least these three names that are easier to understane, i.e. something like `devolatilization_enthalpy_changes`, `water_mass_fractions`, and `devolatilization_onset_temperatures`.

File: source/material_model/reactive_fluid_transport.cc [hunk 15]
```diff
@@ -50,6 +50,96 @@ namespace aspect
...
+          }
+
+      for (unsigned int i = 0; i<Td_all_poly_coeffs.size(); ++i)
+        for (unsigned int j = 0; j<Td_all_poly_coeffs[i].size(); ++j)
+          {
+            Td_values[i] += Td_all_poly_coeffs[i][j] * std::pow(pressure, Td_all_poly_coeffs[i].size() - 1 - j);
+          }
+
+      // Create an array for the equilibrium bound water content that is calculated from these polynomials
+      std::vector<double> eq_bound_water_content;
```
Comment (jdannberg) [inline source/material_model/reactive_fluid_transport.cc:125]: Since you already know the size of the vector (it needs to have the same size as LR_values), you can already initiate the vector with the correct size rather than using push_back later on (with push_back, you have to change how much space in memory the vector needs, which is inefficient).

File: source/material_model/reactive_fluid_transport.cc [hunk 16]
```diff
@@ -50,6 +50,96 @@ namespace aspect
...
+          }
+
+      for (unsigned int i = 0; i<csat_all_poly_coeffs.size(); ++i)
+        for (unsigned int j = 0; j<csat_all_poly_coeffs[i].size(); ++j)
+          {
+            csat_values[i] += i==3 ? csat_all_poly_coeffs[i][j] * std::pow(std::log10(pressure), csat_all_poly_coeffs[i].size() - 1 - j) :\
+                              csat_all_poly_coeffs[i][j] * std::pow(pressure, csat_all_poly_coeffs[i].size() - 1 - j);
+          }
+
+      for (unsigned int i = 0; i<Td_all_poly_coeffs.size(); ++i)
```
Comment (jdannberg) [inline source/material_model/reactive_fluid_transport.cc:118]: As I understand it, this outer loop is the same for all three parameters. (It always has length 4). 
So you can make the code shorter by doing the outer loop only once:
```suggestion
      for (unsigned int i = 0; i<LR_all_poly_coeffs.size(); ++i)
      {
        for (unsigned int j = 0; j<LR_all_poly_coeffs[i].size(); ++j)
          {
            LR_values[i] += LR_all_poly_coeffs[i][j] * std::pow(inverse_pressure, LR_all_poly_coeffs[i].size() - 1 - j);
          }

        for (unsigned int j = 0; j<csat_all_poly_coeffs[i].size(); ++j)
          {
            csat_values[i] += i==3 ? csat_all_poly_coeffs[i][j] * std::pow(std::log10(pressure), csat_all_poly_coeffs[i].size() - 1 - j) :\
                              csat_all_poly_coeffs[i][j] * std::pow(pressure, csat_all_poly_coeffs[i].size() - 1 - j);
          }

```

File: source/material_model/reactive_fluid_transport.cc [hunk 17]
```diff
@@ -75,18 +165,44 @@ namespace aspect
...
+                const unsigned int gabbro_idx = this->introspection().compositional_index_for_name("gabbro");
+                const unsigned int peridotite_idx = this->introspection().compositional_index_for_name("peridotite");
+
+                // Initialize a vector that stores the compositions (mass fractions) for
+                // the four different rock compositions,
+                std::vector<double> tracked_rock_mass_fractions;
+                tracked_rock_mass_fractions.push_back(in.composition[q][peridotite_idx]);
+                tracked_rock_mass_fractions.push_back(in.composition[q][gabbro_idx]);
+                tracked_rock_mass_fractions.push_back(in.composition[q][MORB_idx]);
+                tracked_rock_mass_fractions.push_back(in.composition[q][sediment_idx]);
```
Comment (jdannberg) [inline source/material_model/reactive_fluid_transport.cc:187]: Same here: Don't use push-back, but create the vector with the right size and fill it (or just initialize it with the right values).

File: source/material_model/reactive_fluid_transport.cc [hunk 18]
```diff
@@ -75,18 +165,44 @@ namespace aspect
...
+
+                // The bound water content (water within the solid phase) for the four different rock types
+                std::vector<double> tian_eq_bound_water_content = tian_equilibrium_bound_water_content(in, q);
+
+                // average the water content between the four different rock types
+                double average_eq_bound_water_content = MaterialUtilities::average_value (tracked_rock_mass_fractions, tian_eq_bound_water_content, MaterialUtilities::arithmetic);
+
+                // The fluid volume fraction (stored in the melt_fractions vector) is equal to the porosity and the change
+                // in bound fluid content (current bound fluid - updated average bound fluid).
```
Comment (jdannberg) [inline source/material_model/reactive_fluid_transport.cc:196]:
```suggestion
                // The fluid volume fraction in equilibrium with the solid (stored in the melt_fractions vector) is equal to the sum of the porosity and the change
                // in bound fluid content (current bound fluid - updated average bound fluid).
```

Comment (jdannberg) [reply source/material_model/reactive_fluid_transport.cc:196]: You can also describe this differently (which I would maybe find more intuitive):
The total water fraction is the sum of the current bound water fraction and the porosity. To get the free water fraction in equilibrium with the solid, we have to take this total water fraction and subtract the equilibrium bound water fraction from the Tian parametrization.

File: source/material_model/reactive_fluid_transport.cc [hunk 19]
```diff
@@ -246,12 +362,26 @@ namespace aspect
...
+                             "The maximum allowed weight percent that the sediment composition can hold.");
+          prm.declare_entry ("Maximum weight percent water in MORB", "2",
+                             Patterns::Double (0),
+                             "The maximum allowed weight percent that the sediment composition can hold.");
+          prm.declare_entry ("Maximum weight percent water in gabbro", "1",
+                             Patterns::Double (0),
+                             "The maximum allowed weight percent that the sediment composition can hold.");
+          prm.declare_entry ("Maximum weight percent water in peridotite", "8",
+                             Patterns::Double (0),
+                             "The maximum allowed weight percent that the sediment composition can hold.");
```
Comment (jdannberg) [inline source/material_model/reactive_fluid_transport.cc:376]: I have a question about these parameters. Is there a reason these are input parameters rather than part of the parametrization itself (like all the other parameters)? Or in other words: Are these related to physical properties of the rocks, or do they come from somewhere else (in which case, I would make a note of that).

**Comment by naliboff** [comment]
@danieldouglas92 - To follow up on a comment from my last review, I think it would be really useful to include some figures in the discussion to show how running the tests at higher resolutions can be used to reproduce the figures in the Tian paper. You could also include the python script showing how to make the plots, and then later add it is as part of a new cookbook.

File: source/material_model/reactive_fluid_transport.cc [hunk 20]
```diff
@@ -281,6 +411,11 @@ namespace aspect
           fluid_compressibility             = prm.get_double ("Fluid compressibility");
           fluid_reaction_time_scale         = prm.get_double ("Fluid reaction time scale for operator splitting");
 
+          tian_max_peridotite_water              = prm.get_double ("Maximum weight percent water in peridotite");
+          tian_max_gabbro_water                  = prm.get_double ("Maximum weight percent water in gabbro");
+          tian_max_MORB_water                    = prm.get_double ("Maximum weight percent water in MORB");
+          tian_max_sediment_water                = prm.get_double ("Maximum weight percent water in sediment");
```
Comment (jdannberg) [inline source/material_model/reactive_fluid_transport.cc:417]: Can you align the `= prm.get_double` with the lines above?

File: tests/tian_peridotite.prm [hunk 21]
```diff
@@ -0,0 +1,125 @@
+# This model generates a representation of a phase diagram for the equilibrium
+# wt% water that can be stored in an upper mantle peridotite (PUM) as shown in
+# Figure 11 of  Tian et. al., 2019. This model reproduces that phase diagram
+# (very low res by default), with an important caveat being that the pressure axis
+# is inverted relative to the ASPECT output. A surface pressure of 600 MPa is imposed
+# to reproduce the pressure axis,and a lateral temperature gradient is imposed to
+# reproduce the temperature axis.
+set Nonlinear solver scheme                = iterated Advection and Stokes
+set Max nonlinear iterations               = 10
+set Nonlinear solver tolerance             = 1
```
Comment (jdannberg) [inline tests/tian_peridotite.prm:10]: Are you sure this is what you want?

**Review by jdannberg — COMMENTED** [review]
Thank you for adding this to ASPECT. I've made some suggestions, but otherwise this looks good to me!

File: source/material_model/reactive_fluid_transport.cc (hunk 19 above)
Comment (danieldouglas92) [reply source/material_model/reactive_fluid_transport.cc:376]: They come from somewhere else, the polynomial can reach very high values if it's unbounded at low P-T conditions, so this helps to keep the wt%'s reasonable! I made it variable because I figured a user might want to adjust this depending on what they think is going on in the system their measuring.

File: tests/tian_peridotite.prm (hunk 21 above)
Comment (danieldouglas92) [reply tests/tian_peridotite.prm:10]: Definitely not!! Thanks for catching that

File: include/aspect/material_model/reactive_fluid_transport.h [hunk 22]
```diff
@@ -62,6 +62,19 @@ namespace aspect
...
 
+
+        /**
+         * Compute the maximum allowed bound water content at the input
+         * pressure and temperature conditions. This is used to determine
+         * how free water interacts with the solid phase.
+         * @param in Object that contains the current conditions.
+         * @param q unsigned int from 0-3 indexing which rock phase the equilbrium
+         * bound water content is being calculated for
+         * allowable free fluid fraction for each given input conditions.
```
Comment (jdannberg) [inline include/aspect/material_model/reactive_fluid_transport.h:73]: This last line seems unrelated?

File: include/aspect/material_model/reactive_fluid_transport.h [hunk 23]
```diff
@@ -164,12 +217,26 @@ namespace aspect
          * base model to be incompressible, as otherwise the advection
          * equation would only be valid for mass and not volume
          * fractions.
+         *
+         * The tian approximation model implements parametrized phase
+         * diagrams from Tian et al., 2019 G3, https://doi.org/10.1029/2019GC008488
+         * and calculates the fluid-solid reactions for for four different rock types:
```
Comment (jdannberg) [inline include/aspect/material_model/reactive_fluid_transport.h:223]:
```suggestion
         * and calculates the fluid-solid reactions for four different rock types:
```

**Review by jdannberg — APPROVED** [review]
Looks good to me (other than the two typos)!

**Comment by jdannberg** [comment]
But I agree with @naliboff, it would be great to include an output of one of the tests in the discussion in this pull request, just so that we have a documentation of how it looks like in relation to the figures from the Tian model.

**Comment by danieldouglas92** [comment]
> But I agree with @naliboff, it would be great to include an output of one of the tests in the discussion in this pull request, just so that we have a documentation of how it looks like in relation to the figures from the Tian model.

Updated it in the initial PR message!

**Review by naliboff — APPROVED** [review]
